### ISSUE TITLE
Feature request: UI testing environment

### ISSUE THREAD

**Issue by thibaudcolas** [wagtail/wagtail#5271]
Working on Wagtail’s admin UI, I find it unnecessarily hard to make sure that I take all of the variations in UI components into account. It's a problem with all projects that have fairly big UIs, but is compounded in Wagtail by the fact that so much of the UI is dependent on what content has been created in the CMS, and on how the site has been set up. It also doesn't help that Wagtail's default dev environment is first meant to be a good demo of how Wagtail would be used on a small-scale website, rather than a dev playground (https://github.com/wagtail/bakerydemo/issues/175).

Having a more comprehensive UI "test" environment would help in a number of scenarios:

- Reviewing the existing UI, for example as part of a UX review, accessibility testing, compliance, or just when testing the impact of a specific change.
- Changing existing UI components, and reviewing changes. It's really hard to know what impact a UI change might have, if you have to know how to use the CMS for bits of its UI to appear. This is particularly problematic when multiple people work on the same feature, and they all have to do the same error-prone setup on their local copies of Wagtail for them to test their changes meaningfully
- Setting up automated UI testing in continuous integration – accessibility testing (https://github.com/wagtail/wagtail/issues/4871#issuecomment-486648541), visual regression testing. These are impossible to set up if we don't have a standardised test environment to build the test suites for.

Related: #3804, https://github.com/wagtail/bakerydemo/issues/175, https://github.com/wagtail/wagtail/issues/4871#issuecomment-486648541, https://github.com/wagtail/rfcs/pull/37

---

This was previously discussed at a Wagtail team meeting, and the conclusion there was that it should be done as a separate environment / project, rather than extending `bakerydemo`. It's not entirely clear how this project could then be used in the Wagtail CI builds. Thoughts welcome!

**Comment by jonnyscholes** [comment]
I'm super keen for this too. Working on refactoring some of our older JS, and some of it really could use with some tests.

I'm keen to take a stab at this as part of that refactor as now really is the right time for me to be writing tests whilst I know about the edge cases. Rough thoughts on a path forward:

- Take bakerydemo as a starting point
- Remove vagrant but keeping the docker setup with the intention of running our tests inside docker
- Keep the models as a starting point, but implement testdata in a way that can easily be modified (probably by creating some JSON of just the data, and the importing the models and programatically creating them)

**Comment by thibaudcolas** [comment]
Yep, I’ve also struggled with turning this into an achievable task. My WIP has been to at least tweak the settings to make it easier to automate browsing on the bakery demo: https://github.com/thibaudcolas/bakerydemo/pull/2.

Then my plan was to start making test content for all of the scenarios to support, defined here https://github.com/thibaudcolas/wagtail-tooling/blob/master/ui/scenarios.js (not just individual items within the "scenarios" array, but also all of the states in each of those items).

Once that’s done, all of the cases listed in https://docs.google.com/spreadsheets/d/1rTMilzKhcb43Y3fZKGJJeoKWLCR00bfgwVB9f6OaC2U/edit#gid=0 should be testable.

**Comment by jonnyscholes** [comment]
Yeah sure. The hurdle I've hit with bakerydemo is adding lots of additional
data (without doing it by hand) is hard. There are a few pagination states
that don't get hit with the current data. (Although there may well be an
easy way to do this that I don't know about...)

**Comment by jonnyscholes** [comment]
I've put up a [proof of concept wagtail environment](https://github.com/jonnyscholes/habitat)  thats designed to be used by Wagtail developers and our CI. Yet to see how Travis goes with it, but you can run docker inside Travis so hopefully should be smooth sailing. Not completely happy with how we run the tests, but It Works™️.

Also verrrrry roughly mocked up using `Pa11y`'s actions along side one of your scenarios and that works really well! Loads more to do but it's a start. 

Be interested to get your thoughts on this @thibaudcolas given right now it would probably be mostly you and I using it, if we think it's on the right track I'll look into proposing how we could integrate it with our build pipeline (and start writing some tests for my JS refactoring 🎉).

**Comment by thibaudcolas** [comment]
@marteki interested in this

**Comment by thibaudcolas** [comment]
FYI we briefly discussed this today. I’ve kept https://github.com/thibaudcolas/wagtail-tooling maintained, especially to help with the accessibility work in https://github.com/wagtail/wagtail/projects/5.

At the moment it contains a test suite of a 100+ cases, based on content from the bakerydemo. I regularly use this to review pull requests that change a lot of parts of the UI.

I don’t have much time to spend on this currently but if anyone is interested in using this / picking it up, the tooling is available [in the repository](https://github.com/thibaudcolas/wagtail-tooling). @allcaps was working on a UI testing environment directly within the Wagtail repository during the last sprint in Bristol, which would greatly simplify integrating this into our development workflow (and in CI).

**Comment by thibaudcolas** [comment]
I believe this has been addressed in #7751 so will close this now.
